Iter5

MAGeorg opened the pull request.

Commit:
iter1: add server

Commit:
iter1: change var name

Commit:
iter2

Commit:
iter3

Commit:
iter4: add config with flag

Commit:
iter5

Commit:
iter5: fix run test

arttet commented on internal/config/config.go:
```diff
@@ -0,0 +1,25 @@
+	BaseAddress string
+}
+
+var Conf Config
```
CHANGE REQUEST: please avoid global variables

MAGeorg replied:
я довольно долго думал, не совсем понятно как в таком случае пробрасывать данные из config - BaseAddress в Handler
предлагаете мне переписать стандартный HandleFunc?
```
type HandlerFunc func(ResponseWriter, *Request)

func (f HandlerFunc) ServeHTTP(w ResponseWriter, r *Request) {
	f(w, r)
}
```

arttet commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,64 @@
+	"github.com/MAGeorg/shortener.git/internal/utils"
+)
+
+var savedURL = map[uint32]string{}
```
CHANGE REQUEST: please create a new layer for it

arttet commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,64 @@
+var savedURL = map[uint32]string{}
+
+// генерация сокращенного URL
+func generateURL(s string) uint32 {
```
CHANGE REQUEST: avoid business logic inside controllers

arttet commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,64 @@
+func GetOriginURL(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "text/plain")
+
+	s := r.URL.String()[1:]
```
CHANGE REQUEST: this is an unsafe operation

arttet reviewed (changes requested):
Please integrate the following actions in CI:
- https://about.codecov.io/blog/getting-started-with-code-coverage-for-golang/
- https://github.com/golangci/golangci-lint

Commit:
iter 5: fix the comment of the review

arttet commented on internal/appcontext/appcontext.go:
```diff
@@ -0,0 +1,18 @@
+	"github.com/MAGeorg/shortener.git/internal/storage"
+)
+
+type AppContext struct {
```
SUGGESTION: it would like better if you rename it like `AppData`

arttet commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,62 @@
+	f   func(ctx *appcontext.AppContext, w http.ResponseWriter, r *http.Request)
+}
+
+func (h AppHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
CHANGE REQUEST: please remove it

MAGeorg replied:
без ServeHTTP не будет реализации [Handler](https://pkg.go.dev/net/http#Handler)

arttet commented on internal/handlers/handlers.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+// обработка POST запроса
+func CreateHashURL(ctx *appcontext.AppContext, w http.ResponseWriter, r *http.Request) {
```
```suggestion
func (h *AppHandler) CreateHashURL(w http.ResponseWriter, r *http.Request) {
```

arttet commented on internal/handlers/server.go:
```diff
@@ -0,0 +1,21 @@
+func RunServer(ctx *appcontext.AppContext) {
+	r := chi.NewRouter()
+
+	r.Method("POST", "/", AppHandler{ctx, CreateHashURL})
```
```suggestion
	r.Method("POST", "/", h.CreateHashURL)
```

MAGeorg replied:
не скомпилится (причина в комментарии к CHANGE REQUEST)

arttet replied:
This is the example how you need to change it

arttet commented on internal/handlers/server.go:
```diff
@@ -0,0 +1,21 @@
+	r := chi.NewRouter()
+
+	r.Method("POST", "/", AppHandler{ctx, CreateHashURL})
+	r.Method("GET", "/{id}", AppHandler{ctx, GetOriginURL})
```
```suggestion
	r.Method("GET", "/{id}", h.GetOriginURL)
```

arttet commented on internal/handlers/server.go:
```diff
@@ -0,0 +1,21 @@
+
+	log.Printf("Server fun on %s address ...", ctx.Cfg.Address)
+	if err := http.ListenAndServe(ctx.Cfg.Address, r); err != nil {
+		panic(err)
```
CHANGE REQUEST: please return the error

Commit:
Iter5: fix comments
>
> 1. Rename AppContext to AppData
> 2. Add return error in server.RunServer()
3. Add test for packet util.go
4. Add coveragetest and golangci-lint
5. Change *chi.Mux.Method to *chi.Mux.Post

p.s. *chi.Mux.Method expects http.Handler as the third parameter, for this it was necessary to override ServeHTTP, *chi.Mux.Post expects HandlerFunc as the second parameter, so we had disagreements

Commit:
Iter5: fix error from golangci-lint
